import os, glob, sys
from snakemake.utils import R

SAMPLEMAPPING = 'sample_mapping.tsv'
FASTQDIR = 'sra/'
OUTDIR   = 'out/'
TMPDIR   = 'tmp/'

ORGANISM = 'H_sapiens_hg19'

CLIPTRIMOUT = FASTQDIR
REMOVEPCRDUBLICATESOUT = OUTDIR + 'removed_pcr_dub/'
MPILEUPIN = REMOVEPCRDUBLICATESOUT

include: '../../snake/wes/wes_snake.py'

SAMPLENAMES = getSampleNames()
NORMALTUMORMATCHING = getNormalTumorFiles()

rule SRA051153:
    input:
        expand(REMOVEPCRDUBLICATESOUT + '{sample}.bam_stats/report.pdf', sample = SAMPLENAMES),
        expand(VARSCANSOMATICOUT + '{tumorNormalMatching}.snp.vcf', tumorNormalMatching = NORMALTUMORMATCHING)
        #VARSCANSOMATICOUT + 'PCL-016_vs_PCL-016_CTRL.snp.vcf',
        #VARSCANSOMATICOUT + 'PCL-016_vs_PCL-016_CTRL.indel.vcf',
        #VARSCANSOMATICOUT + 'PCL-019_vs_PCL-019_CTRL.snp.vcf',
        #VARSCANSOMATICOUT + 'PCL-019_vs_PCL-019_CTRL.indel.vcf',
        #VARSCANSOMATICOUT + 'PCL-026_vs_PCL-026_CTRL.snp.vcf',
        #VARSCANSOMATICOUT + 'PCL-026_vs_PCL-026_CTRL.indel.vcf'
